File: token-client-core/src/main/java/no/nav/security/token/support/client/core/http/OAuth2HttpHeaders.java
package no.nav.security.token.support.client.core.http;

import java.util.*;

import static java.lang.String.CASE_INSENSITIVE_ORDER;

public class OAuth2HttpHeaders {

    private final Map<String, List<String>> headers;

    private OAuth2HttpHeaders(final Map<String, List<String>> headers) {
        this.headers = Optional.ofNullable(headers).orElse(Map.of());
    }

    public static OAuth2HttpHeaders of(Map<String, List<String>> headers) {
        return new OAuth2HttpHeaders(headers);
    }

    public static Builder builder() {
        return new Builder();
    }

    public Map<String, List<String>> headers() {
        return headers;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        var that = (OAuth2HttpHeaders) o;
        return Objects.equals(headers, that.headers);
    }

    @Override
    public int hashCode() {
        return Objects.hash(headers);
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + " [headers=" + headers + "]";
    }

    public static class Builder {
        private final TreeMap<String, List<String>> headersMap;

        public Builder() {
            headersMap = new TreeMap<>(CASE_INSENSITIVE_ORDER);
        }

        public Builder header(String name, String value) {
            headersMap.computeIfAbsent(name, k -> new ArrayList<>(1))
                .add(value);
            return this;
        }

        public OAuth2HttpHeaders build() {
            return OAuth2HttpHeaders.of(headersMap);
        }
    }

}

File: token-client-core/src/main/kotlin/no/nav/security/token/support/client/core/http/OAuth2HttpHeaders.kt
package no.nav.security.token.support.client.core.http

import java.lang.String.CASE_INSENSITIVE_ORDER
import java.util.*

class OAuth2HttpHeaders (val headers : Map<String,String>) {

    override fun equals(other : Any?) : Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as OAuth2HttpHeaders
        return headers == that.headers
    }

    override fun hashCode() = Objects.hash(headers)

    override fun toString() = "${javaClass.getSimpleName()} [headers=$headers]"

    class Builder(private val headers : TreeMap<String, String> =  TreeMap(CASE_INSENSITIVE_ORDER)) {

        fun header(name : String, value : String) = this.also { headers.computeIfAbsent(name) { value }
        }
        fun build() = of(headers)
    }

    companion object {

        @JvmField
        val NONE = OAuth2HttpHeaders(emptyMap())
       @JvmStatic
        fun of(headers : Map<String, String>) = OAuth2HttpHeaders(headers)

        @JvmStatic
        fun builder() = Builder()
    }
}